subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/deps/ClassSetAnalysis.java

```diff
@@ -244,11 +205,28 @@ private void processTransitiveDependentClasses(Set<String> visitedClasses,
             }
             accessibleResultClasses.add(accessibleDependentClass);
             DependentsSet currentDependents = getDependents(accessibleDependentClass);
-            if (!currentDependents.isDependencyToAll()) {
+            if (currentDependents.isDependencyToAll()) {
+                return currentDependents.getDescription();
+            } else {
                 resultResources.addAll(currentDependents.getDependentResources());
+                privateResultClasses.addAll(currentDependents.getPrivateDependentClasses());
```

Comment:
These were lost before and we even had a unit test that asserted this loss of information. After removing all the other special-case code, I could confirm that at least one integration test fails if this line is not here. I've corrected the unit test to assert the right behavior.